src/awkward/_v2/contents/listoffsetarray.py

```diff
@@ -570,6 +570,220 @@ def _localindex(self, axis, depth):
+                raise ValueError(
+                    "array with strings can only be checked on uniqueness with axis=-1"
+                )
+
+            # FIXME: check validity error
+
+            if isinstance(self._content, ak._v2.contents.NumpyArray):
+                out, outoffsets = self._content._as_unique_strings(self._offsets)
+                out2 = ak._v2.contents.listoffsetarray.ListOffsetArray(
+                    outoffsets,
+                    out,
+                    None,
+                    self._parameters,
+                )
+                return len(out2) == len(self)
+
+        if negaxis is None:
+            return self._content._is_unique(negaxis, starts, parents, outlength)
+
+        if not branch and (negaxis == depth):
+            return self._content._is_unique(negaxis - 1, starts, parents, outlength)
+        else:
+            nplike = self.nplike
+
+            nextparents = ak._v2.index.Index64.empty(
+                self._offsets[-1] - self._offsets[0], nplike
+            )
+
+            self._handle_error(
+                nplike[
+                    "awkward_ListOffsetArray_reduce_local_nextparents_64",
+                    nextparents.dtype.type,
+                    self._offsets.dtype.type,
+                ](
+                    nextparents.to(nplike),
+                    self._offsets.to(nplike),
+                    len(self._offsets) - 1,
+                )
+            )
+            starts = self._offsets[:-1]
+
+            return self._content._is_unique(negaxis, starts, nextparents, outlength)
+
+    def _unique(self, negaxis, starts, parents, outlength):
+        nplike = self.nplike
+        if len(self._offsets) - 1 == 0:
+            return self
+
+        branch, depth = self.branch_depth
+
+        if (
+            self.parameter("__array__") == "string"
+            or self.parameter("__array__") == "bytestring"
+        ):
+            if branch or (negaxis != depth):
+                raise ValueError("array with strings can only be sorted with axis=-1")
+
+            # FIXME: check validity error
+
+            if isinstance(self._content, ak._v2.contents.NumpyArray):
+                out, nextoffsets = self._content._as_unique_strings(self._offsets)
+                return ak._v2.contents.ListOffsetArray(
+                    nextoffsets,
+                    out,
+                    None,
+                    self._parameters,
+                )
+
+        if not branch and (negaxis == depth):
+            if (
+                self.parameter("__array__") == "string"
+                or self.parameter("__array__") == "bytestring"
+            ):
+                raise ValueError("array with strings can only be sorted with axis=-1")
+            if len(self._offsets) - 1 != len(parents):
+                raise ValueError(
+                    "length of self._offsets_ - 1 is not equal to parents length"
+                )
```

Comment:
This looks more like it would catch an internal programming error than a user mistake. ValueErrors are for user mistakes (and the error message would have to be more helpful, to tell the user what to do instead) and AssertionErrors are for programming mistakes.

If you're thinking that it really shouldn't be possible to get `len(self._offsets) - 1 != len(parents)`, which would be my expectation, too, you can just do

```python
assert len(self._offsets) - 1 == len(parents)
```

and we'll deal with it if it ever happens in practice. No need to come up with a good error message for an assertion unless if provides some new information—if we ever hit it, we'll be getting into the code, anyway. The same doesn't apply to ValueErrors because user mistakes should not be resolved by delving into the code, but by changing inputs.

Reply:
done